File: go/test/fondo_comun_inversion_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/argentinadatos-sdk/go"
	"github.com/voxgig-sdk/argentinadatos-sdk/go/core"

	vs "github.com/voxgig-sdk/argentinadatos-sdk/go/utility/struct"
)

func TestFondoComunInversionEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.FondoComunInversion(nil)
		if ent == nil {
			t.Fatal("expected non-nil FondoComunInversionEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := fondo_comun_inversionBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "fondo_comun_inversion." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set ARGENTINADATOS_TEST_FONDO_COMUN_INVERSION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		fondoComunInversionRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.fondo_comun_inversion", setup.data)))
		var fondoComunInversionRef01Data map[string]any
		if len(fondoComunInversionRef01DataRaw) > 0 {
			fondoComunInversionRef01Data = core.ToMapAny(fondoComunInversionRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = fondoComunInversionRef01Data

		// LOAD
		fondoComunInversionRef01Ent := client.FondoComunInversion(nil)
		fondoComunInversionRef01MatchDt0 := map[string]any{}
		fondoComunInversionRef01DataDt0Loaded, err := fondoComunInversionRef01Ent.Load(fondoComunInversionRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if fondoComunInversionRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func fondo_comun_inversionBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "fondo_comun_inversion", "FondoComunInversionTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read fondo_comun_inversion test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse fondo_comun_inversion test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"fondo_comun_inversion01", "fondo_comun_inversion02", "fondo_comun_inversion03", "mercado_dinero01", "mercado_dinero02", "mercado_dinero03", "renta_fija01", "renta_fija02", "renta_fija03", "renta_mixta01", "renta_mixta02", "renta_mixta03", "renta_variable01", "renta_variable02", "renta_variable03", "retorno_total01", "retorno_total02", "retorno_total03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("ARGENTINADATOS_TEST_FONDO_COMUN_INVERSION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"ARGENTINADATOS_TEST_FONDO_COMUN_INVERSION_ENTID": idmap,
		"ARGENTINADATOS_TEST_LIVE":      "FALSE",
		"ARGENTINADATOS_TEST_EXPLAIN":   "FALSE",
		"ARGENTINADATOS_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["ARGENTINADATOS_TEST_FONDO_COMUN_INVERSION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["ARGENTINADATOS_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["ARGENTINADATOS_APIKEY"],
			},
			extra,
		})
		client = sdk.NewArgentinadatosSDK(core.ToMapAny(mergedOpts))
	}

	live := env["ARGENTINADATOS_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["ARGENTINADATOS_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: ts/test/entity/fondo_comun_inversion/FondoComunInversionEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { ArgentinadatosSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('FondoComunInversionEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when ARGENTINADATOS_TEST_LIVE=TRUE.
  afterEach(liveDelay('ARGENTINADATOS_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = ArgentinadatosSDK.test()
    const ent = testsdk.FondoComunInversion()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.ARGENTINADATOS_TEST_LIVE
    for (const op of ['load']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'fondo_comun_inversion.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"ccp","req":false,"type":"`$NUMBER`","index$":0},{"active":true,"name":"fecha","req":false,"type":"`$STRING`","index$":1},{"active":true,"name":"fondo","req":false,"type":"`$STRING`","index$":2},{"active":true,"name":"horizonte","req":false,"type":"`$STRING`","index$":3},{"active":true,"name":"patrimonio","req":false,"type":"`$NUMBER`","index$":4},{"active":true,"name":"tipo","req":false,"type":"`$STRING`","index$":5},{"active":true,"name":"vcp","req":false,"type":"`$NUMBER`","index$":6}],"name":"fondo_comun_inversion","op":{"load":{"input":"data","name":"load","points":[{"active":true,"args":{"params":[{"active":true,"example":"ultimo","kind":"param","name":"fecha","orig":"fecha","reqd":true,"type":"`$STRING`","index$":0}]},"contract":{"id":"GET /v1/finanzas/fci/mercadoDinero/{fecha}","json":"{\"operationId\":\"get-finanzas-fci-mercado-dinero-fecha\",\"parameters\":[{\"description\":\"Fecha de consulta\",\"example\":\"ultimo\",\"in\":\"path\",\"name\":\"fecha\",\"required\":true,\"schema\":{\"format\":\"date\",\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"items\":{\"properties\":{\"ccp\":{\"type\":\"number\"},\"fecha\":{\"type\":\"string\"},\"fondo\":{\"type\":\"string\"},\"horizonte\":{\"enum\":[\"corto\",\"medio\",\"largo\"],\"type\":\"string\"},\"patrimonio\":{\"type\":\"number\"},\"tipo\":{\"type\":\"string\"},\"vcp\":{\"type\":\"number\"}},\"title\":\"FondoComunInversion\",\"type\":\"object\"},\"type\":\"array\"}}},\"description\":\"Devuelve una lista de Fondos Comunes de Inversión\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/v1/finanzas/fci/mercadoDinero/{fecha}","segments":[{"lit":"v1"},{"lit":"finanzas"},{"lit":"fci"},{"lit":"mercadoDinero"},{"var":"fecha"}],"select":{"exist":["fecha"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":0},{"active":true,"args":{"params":[{"active":true,"example":"ultimo","kind":"param","name":"fecha","orig":"fecha","reqd":true,"type":"`$STRING`","index$":0}]},"contract":{"id":"GET /v1/finanzas/fci/rentaFija/{fecha}","json":"{\"operationId\":\"get-finanzas-fci-renta-fija-fecha\",\"parameters\":[{\"description\":\"Fecha de consulta\",\"example\":\"ultimo\",\"in\":\"path\",\"name\":\"fecha\",\"required\":true,\"schema\":{\"format\":\"date\",\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"items\":{\"properties\":{\"ccp\":{\"type\":\"number\"},\"fecha\":{\"type\":\"string\"},\"fondo\":{\"type\":\"string\"},\"horizonte\":{\"enum\":[\"corto\",\"medio\",\"largo\"],\"type\":\"string\"},\"patrimonio\":{\"type\":\"number\"},\"tipo\":{\"type\":\"string\"},\"vcp\":{\"type\":\"number\"}},\"title\":\"FondoComunInversion\",\"type\":\"object\"},\"type\":\"array\"}}},\"description\":\"Devuelve una lista de Fondos Comunes de Inversión\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/v1/finanzas/fci/rentaFija/{fecha}","segments":[{"lit":"v1"},{"lit":"finanzas"},{"lit":"fci"},{"lit":"rentaFija"},{"var":"fecha"}],"select":{"exist":["fecha"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":1},{"active":true,"args":{"params":[{"active":true,"example":"ultimo","kind":"param","name":"fecha","orig":"fecha","reqd":true,"type":"`$STRING`","index$":0}]},"contract":{"id":"GET /v1/finanzas/fci/rentaMixta/{fecha}","json":"{\"operationId\":\"get-finanzas-fci-renta-mixta-fecha\",\"parameters\":[{\"description\":\"Fecha de consulta\",\"example\":\"ultimo\",\"in\":\"path\",\"name\":\"fecha\",\"required\":true,\"schema\":{\"format\":\"date\",\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"items\":{\"properties\":{\"ccp\":{\"type\":\"number\"},\"fecha\":{\"type\":\"string\"},\"fondo\":{\"type\":\"string\"},\"horizonte\":{\"enum\":[\"corto\",\"medio\",\"largo\"],\"type\":\"string\"},\"patrimonio\":{\"type\":\"number\"},\"tipo\":{\"type\":\"string\"},\"vcp\":{\"type\":\"number\"}},\"title\":\"FondoComunInversion\",\"type\":\"object\"},\"type\":\"array\"}}},\"description\":\"Devuelve una lista de Fondos Comunes de Inversión\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/v1/finanzas/fci/rentaMixta/{fecha}","segments":[{"lit":"v1"},{"lit":"finanzas"},{"lit":"fci"},{"lit":"rentaMixta"},{"var":"fecha"}],"select":{"exist":["fecha"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":2},{"active":true,"args":{"params":[{"active":true,"example":"ultimo","kind":"param","name":"fecha","orig":"fecha","reqd":true,"type":"`$STRING`","index$":0}]},"contract":{"id":"GET /v1/finanzas/fci/rentaVariable/{fecha}","json":"{\"operationId\":\"get-finanzas-fci-renta-variable-fecha\",\"parameters\":[{\"description\":\"Fecha de consulta\",\"example\":\"ultimo\",\"in\":\"path\",\"name\":\"fecha\",\"required\":true,\"schema\":{\"format\":\"date\",\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"items\":{\"properties\":{\"ccp\":{\"type\":\"number\"},\"fecha\":{\"type\":\"string\"},\"fondo\":{\"type\":\"string\"},\"horizonte\":{\"enum\":[\"corto\",\"medio\",\"largo\"],\"type\":\"string\"},\"patrimonio\":{\"type\":\"number\"},\"tipo\":{\"type\":\"string\"},\"vcp\":{\"type\":\"number\"}},\"title\":\"FondoComunInversion\",\"type\":\"object\"},\"type\":\"array\"}}},\"description\":\"Devuelve una lista de Fondos Comunes de Inversión\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/v1/finanzas/fci/rentaVariable/{fecha}","segments":[{"lit":"v1"},{"lit":"finanzas"},{"lit":"fci"},{"lit":"rentaVariable"},{"var":"fecha"}],"select":{"exist":["fecha"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":3},{"active":true,"args":{"params":[{"active":true,"example":"ultimo","kind":"param","name":"fecha","orig":"fecha","reqd":true,"type":"`$STRING`","index$":0}]},"contract":{"id":"GET /v1/finanzas/fci/retornoTotal/{fecha}","json":"{\"operationId\":\"get-finanzas-fci-retorno-total-fecha\",\"parameters\":[{\"description\":\"Fecha de consulta\",\"example\":\"ultimo\",\"in\":\"path\",\"name\":\"fecha\",\"required\":true,\"schema\":{\"format\":\"date\",\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"items\":{\"properties\":{\"ccp\":{\"type\":\"number\"},\"fecha\":{\"type\":\"string\"},\"fondo\":{\"type\":\"string\"},\"horizonte\":{\"enum\":[\"corto\",\"medio\",\"largo\"],\"type\":\"string\"},\"patrimonio\":{\"type\":\"number\"},\"tipo\":{\"type\":\"string\"},\"vcp\":{\"type\":\"number\"}},\"title\":\"FondoComunInversion\",\"type\":\"object\"},\"type\":\"array\"}}},\"description\":\"Devuelve una lista de Fondos Comunes de Inversión\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/v1/finanzas/fci/retornoTotal/{fecha}","segments":[{"lit":"v1"},{"lit":"finanzas"},{"lit":"fci"},{"lit":"retornoTotal"},{"var":"fecha"}],"select":{"exist":["fecha"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":4}],"key$":"load"}},"relations":{"ancestors":[["mercado_dinero"],["renta_fija"],["renta_mixta"],["renta_variable"],["retorno_total"]]},"key$":"fondo_comun_inversion","name__orig":"fondo_comun_inversion","Name":"FondoComunInversion","name_":"fondo_comun_inversion","name-":"fondo-comun-inversion","NAME":"FONDO_COMUN_INVERSION","index$":11}, {"active":true,"entity":"fondo_comun_inversion","key$":"BasicFondoComunInversionFlow","kind":"basic","name":"BasicFondoComunInversionFlow","param":{},"step":[{"active":true,"data":{},"input":{"ref":"fondo_comun_inversion_ref01","srcdatavar":"fondo_comun_inversion_ref01_data","suffix":"_dt0"},"match":{"id":"fondo_comun_inversion01"},"op":"load","spec":[],"valid":[{"apply":"TextFieldMark","def":{"mark":"Mark01-fondo_comun_inversion_ref01"}}],"index$":0}]}, 'FondoComunInversion')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let fondo_comun_inversion_ref01_data = Object.values(setup.data.existing.fondo_comun_inversion)[0] as any

    // LOAD: skipped — no entity id field and load requires path params.
    // Entity-var is declared here so later flow steps still compile.
    const fondo_comun_inversion_ref01_ent = client.FondoComunInversion()


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/fondo_comun_inversion/FondoComunInversionTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = ArgentinadatosSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['fondo_comun_inversion01','fondo_comun_inversion02','fondo_comun_inversion03','mercado_dinero01','mercado_dinero02','mercado_dinero03','renta_fija01','renta_fija02','renta_fija03','renta_mixta01','renta_mixta02','renta_mixta03','renta_variable01','renta_variable02','renta_variable03','retorno_total01','retorno_total02','retorno_total03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'ARGENTINADATOS_TEST_FONDO_COMUN_INVERSION_ENTID': idmap,
    'ARGENTINADATOS_TEST_LIVE': 'FALSE',
    'ARGENTINADATOS_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['ARGENTINADATOS_TEST_FONDO_COMUN_INVERSION_ENTID']

  const live = 'TRUE' === env.ARGENTINADATOS_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['ARGENTINADATOS_TEST_FONDO_COMUN_INVERSION_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new ArgentinadatosSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.ARGENTINADATOS_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  
